internal/brokers/broker.go

```diff
@@ -305,3 +287,76 @@ func (b Broker) validateUILayout(sessionID string, layout map[string]string) (r
 func (b Broker) parseSessionID(sessionID string) string {
 	return strings.TrimPrefix(sessionID, fmt.Sprintf("%s-", b.ID))
 }
+
+// validateAndMarshalUser checks if the specified rawMsg is a valid userinfo and marshals it as a JSON string.
+func (b Broker) validateAndMarshalUser(rawMsg json.RawMessage) (data string, err error) {
+	defer decorate.OnError(&err, "invalid user information")
+
+	var uInfo struct {
+		cache.UserInfo
+		UUID   string
+		UGID   string
+		Groups []struct {
+			Name string
+			UGID string
+		}
+	}
+	if err := json.Unmarshal(rawMsg, &uInfo); err != nil {
+		return "", fmt.Errorf("message is not JSON formatted: %v", err)
+	}
+
+	// Validate username
+	if uInfo.Name == "" {
+		return "", fmt.Errorf("empty username")
+	}
+
+	// Validate home and shell directories
+	for _, dir := range []string{uInfo.Dir, uInfo.Shell} {
+		if !filepath.IsAbs(filepath.Clean(dir)) {
+			return "", fmt.Errorf("%q is not an absolute path", dir)
+		}
+	}
+
+	// Generate UID and GIDs
+	hasher := crypto.SHA256.New()
+
+	uid, err := hasher.Write([]byte(b.Name + uInfo.UUID))
+	if err != nil {
+		return "", fmt.Errorf("could not generate UID")
+	}
+	uInfo.UID = (uid % (100000 - 65537)) + 65536 // Ensures that UID is between 65536 and 100000
+
+	for _, g := range uInfo.Groups {
+		group := cache.GroupInfo{Name: g.Name}
+
+		gid, err := hasher.Write([]byte(b.Name + g.UGID))
```

Comment:
we probably needs to extract that to a small function called for user and groups.